File: ios/BFExplorer/Plaoc/IPCCore/IPC/IpcBodyReceiver.swift
//
//  IpcBodyReceiver.swift
//  BFExplorer
//
//  Created by ui03 on 2023/2/27.
//

import UIKit

/**
 * metaBody 可能会被多次转发，
 * 但只有第一次得到这个 metaBody 的 ipc 才是它真正意义上的 Receiver
 */

class IpcBodyReceiver: IpcBody {
    
    init(metaBody: MetaBody?, ipc: Ipc?) {
        super.init()
        self.metaBody = metaBody
        self.ipc = ipc
        self.bodyHub = hub
        
        // 将第一次得到这个metaBody的 ipc 保存起来，这个ipc将用于接收
        if metaBody?.isStream ?? false {
            guard metaBody != nil else { return }
            let metaIpc = metaId_receiverIpc_Map[metaBody!.metaId]
            if metaIpc == nil {
                _ = ipc?.onClose(cb: { _ in
                    metaId_receiverIpc_Map.removeValue(forKey: metaBody!.metaId)
                })
                metaBody?.receiverUid = ipc?.uid
                metaId_receiverIpc_Map[metaBody!.metaId] = ipc
            }
        }
    }
    
    lazy private var hub: BodyHub = {
        var body = BodyHub()
//        guard metaBody != nil else { return body }
//        var data: Any?
//        if metaBody!.isStream {
//            guard let ipc = metaId_receiverIpc_Map[metaBody!.metaId] else {
//                print("no found ipc by metaId:\(metaBody!.metaId)")
//                return
//            }
//            metaToStream(metaBody: metaBody, ipc: ipc)
//        } else {
//            if metaBody!.encoding == .UTF8 {
//                if let dataStr = metaBody?.data as? String {
//                    data = dataStr
//                }
//            } else if metaBody!.encoding == .BINARY {
//                if let bytes = metaBody?.data as? [UInt8] {
//                    data = bytes
//                }
//            } else if metaBody!.encoding == .BASE64 {
//                if let dataStr = metaBody?.data as? String {
//                    data = dataStr.fromBase64()
//                }
//            } else {
//                print("invalid metaBody type:\(metaBody!.type)")
//                return
//            }
//        }
//        body.data = data
//        if let dataStr = data as? String {
//            body.text = dataStr
//        } else if let bytes = data as? [UInt8] {
//            body.u8a = bytes
//        } else if let stream = data as? InputStream {
//            body.stream = stream
//        }
        return body
    }()
    
    func from(metaBody: MetaBody, ipc: Ipc) -> IpcBody {
        return metaId_ipcBodySender_Map[metaBody.metaId] ?? IpcBodyReceiver(metaBody: metaBody, ipc: ipc)
    }
    
    func metaToStream(metaBody: MetaBody?, ipc: Ipc?) -> InputStream {
        
        let stream_id = metaBody?.streamId ?? ""
        
        let stream = ReadableStream(cid: "receiver-\(stream_id)") { controller in
            var firstData: [UInt8]?
            if metaBody?.encoding == .UTF8 {
                firstData = (metaBody?.data as? String)?.fromUtf8()
            } else if metaBody?.encoding == .BINARY {
                firstData = metaBody?.data as? [UInt8]
            } else if metaBody?.encoding == .BASE64 {
                firstData = (metaBody?.data as? String)?.fromBase64()
            }
            if firstData != nil {
                controller.enqueue(Data(firstData!))
            }
            _ = ipc?.onMessage(cb: { message,_ in
                if let request = message as? IpcStreamData, request.stream_id == stream_id {
                    controller.enqueue(Data(request.binary ?? []))
                } else if let request = message as? IpcStreamEnd, request.stream_id == stream_id {
                    controller.close()
                }
            })
        } onPull: { desiredSize, controller in
            ipc?.postMessage(message: IpcStreamPull(stream_id: stream_id, desiredSize: 1))
        }

        return stream
    }
}

File: next/dweb-browser/ipc/IpcBodyReceiver.cs
namespace ipc;

/**
 * <summary>
 * metaBody 可能会被多次转发，
 * 但只有第一次得到这个 metaBody 的 ipc 才是它真正意义上的 Receiver
 * </summary>
 */
public class IpcBodyReceiver : IpcBody
{
    public Ipc Ipc { get; set; }
    public IpcBodyReceiver(SMetaBody metaBody, Ipc ipc)
    {
        MetaBody = metaBody;
        Ipc = ipc;

        var ipcMetaBodyType = new SMetaBody.IpcMetaBodyType(MetaBody.Type);
        if (ipcMetaBodyType.IsStream)
        {
            if (!CACHE.MetaId_receiverIpc_Map.TryGetValue(MetaBody.MetaId, out Ipc? cipc))
            {
                Ipc.OnClose += () =>
                {
                    CACHE.MetaId_receiverIpc_Map.Remove(MetaBody.MetaId);
                };

                MetaBody = MetaBody with { ReceiverUid = Ipc.Uid };
                CACHE.MetaId_receiverIpc_Map.AddOrUpdate(MetaBody.MetaId, Ipc);
            }
        }
    }

    protected override BodyHubType BodyHub
    {
        get
        {
            return new Lazy<BodyHubType>(new Func<BodyHubType>(() => new BodyHubType().Also(it =>
            {
                object data = default;
                var ipcMetaBodyType = new SMetaBody.IpcMetaBodyType(MetaBody.Type);

                if (ipcMetaBodyType.IsStream)
                {
                    if (!CACHE.MetaId_receiverIpc_Map.TryGetValue(MetaBody.MetaId, out Ipc? ipc))
                    {
                        throw new Exception($"no found ipc by metaId: {MetaBody.MetaId}");
                    }

                    data = MetaToStream(MetaBody, ipc);
                }
                else
                {
                    switch (ipcMetaBodyType.Encoding)
                    {
                        case IPC_DATA_ENCODING.UTF8:
                            data = (string)MetaBody.Data;
                            break;
                        case IPC_DATA_ENCODING.BINARY:
                            data = (byte[])MetaBody.Data;
                            break;
                        case IPC_DATA_ENCODING.BASE64:
                            data = ((string)MetaBody.Data).FromBase64();
                            break;
                        default:
                            throw new Exception($"invalid metaBody type {MetaBody.Type}");
                    }
                }

                it.Data = data;

                switch (data)
                {
                    case string str:
                        it.Text = str;
                        break;
                    case byte[] byteArray:
                        it.U8a = byteArray;
                        break;
                    case Stream stream:
                        it.Stream = stream;
                        break;
                }
            })), true).Value;
        }
    }

    public override SMetaBody MetaBody { get; set; }

    public override object? Raw
    {
        get { return BodyHub.Data; }
    }

    public static IpcBody From(SMetaBody metaBody, Ipc ipc) =>
        CACHE.MetaId_ipcBodySender_Map.TryGetValue(metaBody.MetaId, out IpcBody? ipcBody) ? ipcBody : new IpcBodyReceiver(metaBody, ipc);

    /**
     * <returns> {String | ByteArray | InputStream} </returns>
     */
    public static Stream MetaToStream(SMetaBody metaBody, Ipc ipc)
    {
        /// metaToStream
        var stream_id = metaBody.StreamId!;
        /**
         * 默认是暂停状态
         */
        var paused = 1;
        var stream = new ReadableStream($"receiver-{stream_id}",
            onStart: controller =>
            {
                /// 如果有初始帧，直接存起来
                var ipcMetaBodyType = new SMetaBody.IpcMetaBodyType(metaBody.Type);

                switch (ipcMetaBodyType.Encoding)
                {
                    case IPC_DATA_ENCODING.UTF8:
                        controller.Enqueue(((string)metaBody.Data).FromUtf8());
                        break;
                    case IPC_DATA_ENCODING.BINARY:
                        controller.Enqueue((byte[])metaBody.Data);
                        break;
                    case IPC_DATA_ENCODING.BASE64:
                        controller.Enqueue(((string)metaBody.Data).FromBase64());
                        break;
                    default:
                        break;
                }


                OnMessageHandler<IpcStream, Ipc> onStream = null!;
                onStream = async (ipcStream, ipc) =>
                {
                    if (ipcStream is IpcStreamData data && data.StreamId == stream_id)
                    {
                        Console.WriteLine($"receiver/StreamData/{ipc}/{controller.Stream}", data);
                        await controller.EnqueueAsync(data.Binary);
                    }
                    else if (ipcStream is IpcStreamEnd end && end.StreamId == stream_id)
                    {
                        Console.WriteLine($"receiver/StreamEnd/{ipc}/{controller.Stream}", end);
                        controller.Close();
                        ipc.OnStreamEvent.Remove(onStream);
                    }

                };

                ipc.OnStreamEvent.Listen(onStream);

                //ipc.OnStream(async (IpcStreamMessageArgs args) =>
                //{
                //    if (args.stream is IpcStreamData data && data.StreamId == stream_id)
                //    {
                //        Console.WriteLine($"receiver/StreamData/{ipc}/{controller.Stream}", data);
                //        await controller.EnqueueAsync(data.Binary);
                //    }
                //    else if (args.stream is IpcStreamEnd end && end.StreamId == stream_id)
                //    {
                //        Console.WriteLine($"receiver/StreamEnd/{ipc}/{controller.Stream}", end);
                //        controller.Close();
                //        return SIGNAL_CTOR.OFF;
                //    }

                //    return null;
                //});
            },
            onPull: async args =>
            {
                var controller = args.Item2;
                Console.WriteLine($"receiver/StreamEnd/{ipc}/{controller.Stream}", stream_id);
                if (Interlocked.CompareExchange(ref paused, 1, 0) == 1)
                {
                    await ipc.PostMessageAsync(new IpcStreamPulling(stream_id));
                }
            },
            onClose: async () =>
            {
                await ipc.PostMessageAsync(new IpcStreamAbort(stream_id));
            }
            ).Stream;

        Console.WriteLine($"receiver/{ipc}/{stream}");

        return stream;
    }
}